File: python/demo17b_reentrant_lock.py
'''
REENTRANT LOCKS (RECURSIVE MUTEXES)
Version B: Solving the problem from version A
'''

import threading



lock = threading.RLock()



def do_task():
    with lock:
        print('First time acquiring the resource')
        with lock:
            print('Second time acquiring the resource')



th = threading.Thread(target=do_task)
th.start()
th.join()

File: cpp/cpp-std/demo17b-reentrant-lock.cpp
/*
REENTRANT LOCKS (RECURSIVE MUTEXES)
Version B: Solving the problem from version A
*/


#include <iostream>
#include <thread>
#include <mutex>
using namespace std;



std::recursive_mutex mut;



void doTask() {
    mut.lock();
    cout << "First time acquiring the resource" << endl;

    mut.lock();
    cout << "Second time acquiring the resource" << endl;

    mut.unlock();
    mut.unlock();
}



void doTaskUsingSyncBlock() {
    using uniquelk = std::unique_lock<std::recursive_mutex>;

    uniquelk lk(mut);
    cout << "First time acquiring the resource" << endl;

    {
        uniquelk lk(mut);
        cout << "Second time acquiring the resource" << endl;
    }
}



int main() {
    auto th = std::thread(&doTask);
    th.join();
    return 0;
}
